pkg/server/container_create.go

```diff
@@ -297,12 +297,33 @@ func (c *criService) volumeMounts(containerRootDir string, criMounts []*runtime.
 }
 
 // runtimeSpec returns a default runtime spec used in cri-containerd.
-func runtimeSpec(id string, opts ...oci.SpecOpts) (*runtimespec.Spec, error) {
+func (c *criService) runtimeSpec(id string, baseSpecFile string, opts ...oci.SpecOpts) (*runtimespec.Spec, error) {
 	// GenerateSpec needs namespace.
 	ctx := ctrdutil.NamespacedContext()
-	spec, err := oci.GenerateSpec(ctx, nil, &containers.Container{ID: id}, opts...)
+	container := &containers.Container{ID: id}
+
+	if baseSpecFile != "" {
+		baseSpec, ok := c.baseOCISpecs[baseSpecFile]
+		if !ok {
+			return nil, errors.Errorf("can't find base OCI spec %q", baseSpecFile)
+		}
+
+		spec := oci.Spec{}
+		if err := util.DeepCopy(&spec, &baseSpec); err != nil {
+			return nil, errors.Wrap(err, "failed to clone OCI spec")
+		}
+
+		if err := oci.ApplyOpts(ctx, nil, container, &spec, opts...); err != nil {
```

Comment:
After doing this you need to make sure that you setup the cgrouppath based on the namespace of context.

```go
				var s oci.Spec
		if err := json.NewDecoder(f).Decode(&s); err != nil {
			return errors.Wrap(err, "decode spec")
		}
		s.Linux.CgroupsPath = filepath.Join("/", ns, c.ID)

		if err := oci.ApplyOpts(ctx, client, c, &s, opts...); err != nil {
			return errors.Wrap(err, "apply opts")
		}
```

Reply:
Good catch, thank you.